pymc3/distributions/discrete.py

```diff
@@ -603,6 +604,105 @@ def logcdf(value, mu):
+    dtype = "int64"
+    _print_name = ("GeneralizedPoisson", "\\operatorname{GeneralizedPoisson}")
+
+    @classmethod
+    def rng_fn(cls, rng, theta, lam, size):
+        U = rng.uniform(size=size)
+        omega = np.exp(-lam)
+        S = np.exp(-theta)
+        P = np.copy(S)
+        vrandom = np.vectorize(random)
+        return vrandom(U, S, P, theta, lam, omega)
+
+    def random(u, s, p, theta, lam, omega):
+        x = 0
+        while u > s:
+            x += 1
+            c = theta - lam + lam * x
+            p = omega * c * (1 + lam / c)**(x - 1) * p / x
+            s += p
+        return x
+
+
+generalized_poisson = GeneralizedPoissonRV()
+
+
+class GeneralizedPoisson(Discrete):
+    R"""
+    Generalized Poisson log-likelihood.
+
+    Often used to model the number of events occurring in a fixed period
+    of time when the times at which events occur are independent.
+    Offers greater flexibility in disperson than the standard Poisson because
+    the mean and variance do not have to be equal.
+
+    The pmf of this distribution is
+
+    .. math:: f(x \mid \theta, \lambda) = 
+                  \frac{\theta (\theta + \lambda x)^{x-1} e^{-\theta - \lambda x}}{x!}
+
+    ========  ======================================
+    Support   :math:`x \in \mathbb{N}_0`
+    Mean      :math:`\frac{\theta}{1 - \lambda}`
+    Variance  :math:`\frac{\theta}{(1 - \lambda)^3}`
+    ========  ======================================
+
+    Parameters
+    ----------
+    theta: float
+        Mean parameter (theta > 0).
+    lam: float
+        Dispersion parameter (max(-1, -theta/4) <= lam <= 1).
+
+    Notes
+    -----
+    When lam = 0, the Generalized Poisson reduces to the standard Poisson with mu = theta.
+    When lam < 0, the mean is greater than the variance (underdispersion).
+    When lam > 0, the mean is less than the variance (overdispersion).
+    """
+    rv_op = generalized_poisson
+
+    @classmethod
+    def dist(cls, theta, lam, *args, **kwargs):
+        theta = at.as_tensor_variable(floatX(theta))
+        lam = at.as_tensor_variable(floatX(lam))
+        return super().dist([theta, lam], *args, **kwargs)
+
+    def logp(value, theta, lam):
+        r"""
+        Calculate log-probability of Generalized Poisson distribution at specified value.
+
+        Parameters
+        ----------
+        value: numeric
+            Value(s) for which log-probability is calculated. If the log probabilities for multiple
+            values are desired the values must be provided in a numpy array or Aesara tensor
+
+        Returns
+        -------
+        TensorVariable
+        """
+        log_prob = bound(np.log(theta) + logpow(theta + lam * value, value - 1)
+                         - (theta + lam * value) - factln(value),
+                         theta >= 0,
+                         -1 <= lam, -theta/4 <= lam, lam <= 1,
+                         value >= 0)
+        # Probability is 1 when theta and value are both 0.
```

Comment:
It says above in the docstrings that theta must be `>0`. We usually prefer to keep the logp formula simple and to not include limiting cases. 

Can we just enforce the parameter domain to be strictly positive or is it important to include this special case?

Reply:
You are right, the bound should actually be `theta > 0` so we can remove this special case.